contrib/tablefunc/sql/tablefunc.sql

```diff
@@ -204,3 +204,248 @@ INSERT INTO connectby_int VALUES(111,11);
 INSERT INTO connectby_int VALUES(1,111);
 -- this should not fail due to recursion detection
 SELECT * FROM connectby('connectby_int', 'keyid', 'parent_keyid', '11', 0, '-') AS t(keyid int, parent_keyid int, level int, branch text);
+
+
+-- test for CTAS and DML operations
+-- from crosstab(text,text)
+CREATE TABLE ct1(
+    row_name varchar(32),
+    extra_col varchar(32),
+    attr varchar(32),
+    value varchar(32)
+)
+DISTRIBUTED BY (row_name);
+\copy ct1 from 'data/ctas.data'
+
+CREATE TABLE ct2
+(
+    row_name varchar(32),
+    extra    varchar(32),
+    cat1     varchar(32),
+    cat2     varchar(32),
+    cat3     varchar(32),
+    cat4     varchar(32)
+) DISTRIBUTED BY (row_name);
+
+INSERT INTO ct2
+SELECT ct.*
+FROM crosstab(
+  'select row_name, extra_col, attr, value
+  from ct1
+  order by 1',
+  'select ''att'' || seq from generate_series(1, 4) seq order by 1')
+AS ct(row_name varchar, extra varchar, att1 varchar, att2 varchar, att3 varchar, att4 varchar);
+
+SELECT * FROM ct2;
+
+CREATE TABLE ct2_2 AS SELECT ct.*
+FROM crosstab(
+  'select row_name, extra_col, attr, value
+  from ct1
+  order by 1',
+  'select ''att'' || seq from generate_series(1, 4) seq order by 1')
+AS ct(row_name varchar, extra varchar, att1 varchar, att2 varchar, att3 varchar, att4 varchar);
+
+SELECT * FROM ct2_2;
+
+-- from crosstab(text)
+CREATE TABLE ct3 (
+    row_name varchar(32),
+    cat1     varchar(32),
+    cat2     varchar(32),
+    cat3     varchar(32)
+) DISTRIBUTED BY (row_name);
+INSERT INTO ct3
+SELECT *
+FROM crosstab(
+  'select row_name, attr, value
+   from ct1
+   where attr = ''att2'' or attr = ''att3''
+   order by 1,2')
+AS ct(row_name varchar, att1 varchar, att2 varchar, att3 varchar);
+
+SELECT * FROM ct3;
+
+CREATE TABLE ct3_2 AS SELECT *
+FROM crosstab(
+  'select row_name, attr, value
+   from ct1
+   where attr = ''att2'' or attr = ''att3''
+   order by 1,2')
+AS ct(row_name varchar, att1 varchar, att2 varchar, att3 varchar);
+
+SELECT * FROM ct3_2;
+
+-- from crosstab3(text)
+CREATE TABLE ct4(
+    row_name text,
+    extra_col text,
+    attr text,
+    value text
+)
+DISTRIBUTED BY (row_name);
+\copy ct4 from 'data/ctas.data'
```

Comment:
Why did you change `varchar(32)` to `text` in the `crosstab2..4` tests? Can we save previous tests style, use ct1 in these tests and remove ct4?

Reply:
No, these functions return custom types `tablefunc_crosstab_2..4` which have `text`, so the output types have to match. Meanwhile the plain `crosstab` is generic so it returns `record` and type conversion happens inside `AS ct` clause (which isn't supported for custom types). Changing types to `text` is the easiest solution